chainer/variable.py

```diff
@@ -90,6 +90,24 @@ def label(self):
         return '%s, %s' % (str(self.data.shape),
                            str(self.data.dtype))
 
+    @property
+    def grad(self):
+        return self._grad
+
+    @grad.setter
+    def grad(self, g):
+        if g is not None:
+            if not isinstance(g, type(self.data)):
+                raise TypeError('Type of data and grad mismatch: %s != %s'
```

Comment:
It depends on the case. If the user manually sets the grad and its type is wrong, then the user should fix this problem. If there is a function returns a grad with the wrong type, it's a bug. The user may distinguish these situations by watching the stack trace, though is a bit complicated task. I have no idea to distinguish them at this line.